lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp

```diff
@@ -0,0 +1,294 @@
+    ModuleListProperties &properties =
+        ModuleList::GetGlobalModuleListProperties();
+    policy.Interval = std::chrono::hours(1);
+    policy.MaxSizeBytes = properties.GetSwiftMetadataCacheMaxByteSize();
+    policy.Expiration = std::chrono::hours(
+        properties.GetSwiftMetadataCacheExpirationDays() * 24);
+    m_data_file_cache.emplace(ModuleList::GetGlobalModuleListProperties()
+                                  .GetSwiftMetadataCachePath()
+                                  .GetPath(),
+                              policy);
+  }
+}
+
+SwiftMetadataCache *SwiftMetadataCache::get() {
+  if (!ModuleList::GetGlobalModuleListProperties()
+          .GetEnableSwiftMetadataCache()) 
+    return nullptr;
+
+  static SwiftMetadataCache instance;
+  return &instance;
+}
+
+void SwiftMetadataCache::registerModuleWithReflectionInfoID(ModuleSP module,
+                                                            uint64_t info_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  Log *log = GetLog(LLDBLog::Types);
+  auto it = m_info_to_module.find(info_id);
+
+  // We might've registered and loaded the cache for this module already. Return
+  // early in that case.
+  if (it != m_info_to_module.end()) {
+    if (std::get<bool>(it->second)) {
+      LLDB_LOGV(
+          log,
+          "[SwiftMetadataCache] Module {0} already register, returning early.",
+          module->GetFileSpec().GetFilename());
+      return;
+    }
+  }
+
+  m_info_to_module[info_id] = {module, false};
+  loadCacheForModule(module, info_id);
+}
+
+void SwiftMetadataCache::loadCacheForModule(ModuleSP module, uint64_t info_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+
+  Log *log = GetLog(LLDBLog::Types);
+
+  if (!m_data_file_cache)
+    return;
+
+  auto module_name = getTyperefCacheFileNameForModule(module);
+  auto mem_buffer_up = m_data_file_cache->GetCachedData(module_name);
+
+  // Nothing cached.
+  if (!mem_buffer_up) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] No cached file found for module {0}.",
+              module->GetFileSpec().GetFilename());
+    return;
+  }
+
+  // Extractor used to extract the header information (see the .h file for
+  // details on the format).
+  DataExtractor header_extractor(mem_buffer_up->getBufferStart(),
+                                 mem_buffer_up->getBufferSize(),
+                                 module->GetObjectFile()->GetByteOrder(),
+                                 module->GetObjectFile()->GetAddressByteSize());
+
+  lldb::offset_t read_offset = 0;
+
+  std::string UUID = module->GetUUID().GetAsString();
+  std::string cached_UUID = header_extractor.GetCStr(&read_offset);
+  // If no uuid in the file something is wrong with the cache.
+  if (cached_UUID.empty()) {
+    LLDB_LOG(log,
+             "[SwiftMetadataCache] Failed to read cached UUID for module {0}.",
+             module->GetFileSpec().GetFilename());
+    m_data_file_cache->RemoveCacheFile(module_name);
```

Comment:
I would have expected RemoveCacheFile() to be called immediately before a new file is written. What's the flow if the cache is stale?
loadCacheForModule() deletes the cache, but where is a new one written?

Reply:
The new one is written in `registerFieldDescriptors`, which is when we have new data to cache. I think it makes sense to delete the cache immediately if it's stale/corrupted. Since we're including the lldb version as part of the hash now I don't think there would be any advantages in doing that later. Doing it earlier is also nice because we clear up disk space.